### PR TITLE
create an interface for gh creds provider

### PR DESCRIPTION
**Pull request by punkle**
We plan to build on this later to allow the credentials provider to
be passed into the scaffolder tasks, the processors, and the url
readers.

Signed-off-by: Brian Fletcher <brian@roadie.io>

### CODE REVIEW

File: packages/integration/src/github/DefaultGithubCredentialsProvider.ts [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+
+/**
+ * Handles the creation and caching of credentials for GitHub integrations.
+ *
+ * @public
+ * @remarks
+ *
+ * TODO: Possibly move this to a backend only package so that it's not used in the frontend by mistake
+ */
+export class DefaultGithubCredentialsProvider
```
Comment (Rugvip) [inline packages/integration/src/github/DefaultGithubCredentialsProvider.ts:231]: I'm a bit worried about calling this one `Default*`, as it's probably going to have a more specific purpose. Something like `SingleInstanceGithubCredentialsProvider` although that's ofc a bit long. Could potentially also go the other way around by providing a couple of different factory methods for a simple place, like the `UrlReaders`, so that would be something like `GithubCredentialsProvider.createForSingleInstance()` instead

**Review by Rugvip — COMMENTED** [review]
Thanks! 👍

File: packages/integration/src/github/DefaultGithubCredentialsProvider.ts (hunk 1 above)
Comment (punkle) [reply packages/integration/src/github/DefaultGithubCredentialsProvider.ts:231]: Hey @Rugvip.

I am totally happy to update the name of the class. Maybe AppConfigGithubCredentialsProvider might work also?

I think I am not quite understanding the factory method comment. Are you referring to the "ReaderFactory" static functions that URLReaders implement. Is that right?

Am I miss-understanding?

File: packages/integration/src/github/GithubCredentialsProvider.ts [hunk 2]
```diff
@@ -234,63 +33,11 @@ export type GithubCredentials = {
...
-      token = this.token;
-    }
-
-    return {
-      headers: token ? { Authorization: `Bearer ${token}` } : undefined,
-      token,
-      type,
-    };
-  }
+export interface GithubCredentialsProvider {
```
[the remark this answers is not in the document]
Comment (punkle) [reply packages/integration/src/github/GithubCredentialsProvider.ts:41]: totally.

File: packages/integration/src/github/DefaultGithubCredentialsProvider.ts (hunk 1 above)
Comment (punkle) [reply packages/integration/src/github/DefaultGithubCredentialsProvider.ts]: In the meantime, Ive changed the name to `SingleInstanceGithubCredentialsProvider` and added a new type for the static create function.. kind of like the `ReaderFactory` I think.

File: packages/integration/src/github/types.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  getCredentials(opts: { url: string }): Promise<GithubCredentials>;
+}
+
+/**
+ * This allows implementations to provide factories to create credential providers
+ *
+ * @public
+ *
+ */
+export type GithubCredentialsProviderFactory = (
```
Comment (freben) [inline packages/integration/src/github/types.ts:53]: So this one isn't being used yet right?

**Review by freben — COMMENTED** [review]
Looks good to me, except that it needs a changeset. @Rugvip happy too?

File: packages/integration/src/github/types.ts (hunk 3 above)
Comment (punkle) [reply packages/integration/src/github/types.ts:53]: Its used in one place, here https://github.com/backstage/backstage/pull/8533/files#diff-f5728016589f573089e805f0ab0026407d1c1458c3c63f4d6dbe1fc7bcb5412cR235

Comment (punkle) [reply packages/integration/src/github/types.ts:53]: done.

**Comment by punkle** [comment]
added a changeset

File: packages/integration/src/github/types.ts (hunk 3 above)
Comment (Rugvip) [reply packages/integration/src/github/types.ts:53]: Ah, let's not have a special type for that though, can just have the `create` method be a regular static method definition that returns a `SingleInstanceGithubCredentialsProvider` instance

File: .changeset/five-cameras-hammer.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+---
+'@backstage/backend-common': patch
+'@backstage/integration': patch
```
Comment (Rugvip) [inline .changeset/five-cameras-hammer.md:3]: It's a breaking change for `integration`, I think the best thing to do here is to create a separate changeset for just the `integration` package, along with some pointers for how to migrate

**Review by Rugvip — COMMENTED** [review]
Thanks! just a few nits and it's good to go imo

File: packages/integration/src/github/types.ts (hunk 3 above)
Comment (punkle) [reply packages/integration/src/github/types.ts]: Done.

File: .changeset/five-cameras-hammer.md (hunk 4 above)
Comment (punkle) [reply .changeset/five-cameras-hammer.md]: Thanks, done, Ive set it to minor. Hopefully that is correct.

**Review by Rugvip — APPROVED** [review]
Looks good, thanks! 👍 

Let's :shipit:
